### PR TITLE
[RNMobile] Audio block I

### PR DESCRIPTION
**Pull request by etoledom**
`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2854
`WPiOS` PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/15420
`WPAndroid` PR: https://github.com/wordpress-mobile/WordPress-Android/pull/13808

## Description

This PR implements the base for the Audio block.

**Done:**
- Placeholder.
- Upload from device.
- Add from WPMediaLibrary.
- Replace existing file.
- (Basic) Caption field.
- Settings:
  - Autoplay
  - Loop
  - Preload
- Proper Caption Filed (use `BlockCaption` component): https://github.com/WordPress/gutenberg/pull/27689/

**In Progress:**
- Audio Player UI: https://github.com/WordPress/gutenberg/pull/27467
- Add from URL: https://github.com/WordPress/gutenberg/pull/27817
- Improve Upload and Error state.

**To Do:**
- WPAndroid audio block upload processor

## How has this been tested?

**(On iOS for now)**
- Add an Audio block.
- Tap on `ADD AUDIO`
  - Choose `WordPress Media Library`:
    - An audio file should load instantly.
  - Choose Other Apps:
  - Pick an audio file from your device.
    - A mock upload should run showing the progress indicator
- Smoke test the block settings.

## Screenshots 

audio_block_I

## Types of changes

New mobile block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by guarani — COMMENTED** [review]
Sorry @etoledom I forgot to hit send on this from last Friday. Just minor comments (I'm not approving since I wasn't added as a reviewer here, just to https://github.com/wordpress-mobile/WordPress-iOS/pull/15420).

File: packages/block-library/src/audio/edit.native.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+	return (
+		<>
+			<InspectorControls>
+				<PanelBody title={ __( 'Audio settings' ) }>
+					<ToggleControl
+						label={ __( 'Autoplay' ) }
+						onChange={ toggleAttribute( 'autoplay' ) }
+						checked={ autoplay }
+						help={ getAutoplayHelp }
```
[the remark this answers is not in the document]
Comment (etoledom) [reply packages/block-library/src/audio/edit.native.js:171]: This comes from Web, it's not really used on Mobile, I guess we can remove it 👍

**Comment by etoledom** [comment]
@Guarani - Thanks for the review. I guess you are reviewer now! 🎉 😄 

I agree with the changes requested. @jd-alexander could you please take over this PR? Thanks! 🙏

**Comment by jd-alexander** [comment]
> @Guarani - Thanks for the review. I guess you are reviewer now! 🎉 😄 
> 
> 
> 
> I agree with the changes requested. @jd-alexander could you please take over this PR? Thanks! 🙏 

Of course I can @etoledom 😄

File: packages/block-library/src/audio/edit.native.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+			</View>
+		);
+	}
+
+	function getBlockControls( open ) {
+		return (
+			<BlockControls>
+				<ToolbarGroup>
+					<ToolbarButton
+						title={ __( 'Edit audio' ) }
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply packages/block-library/src/audio/edit.native.js]: Resolved here 9e03e63

File: packages/block-library/src/audio/edit.native.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply packages/block-library/src/audio/edit.native.js]: Done 349aa19 👍

**Comment by jd-alexander** [comment]
To the Android reviewer, the changes here might be a lot but most of it has been reviewed including the caption PR (https://github.com/WordPress/gutenberg/pull/27689/files) that was merged into this has already been reviewed. The primary need here is to verify the logic changes in the Java/Android code that's related to supporting the Audio Block. To test the overall behavior of these changes in action, use this [WP Android PR](https://github.com/wordpress-mobile/WordPress-Android/pull/13808). Once those changes are verified then this PR and the associated PRs can be merged in.

**Comment by cameronvoell** [comment]
Hey Joel. The Android associated code all looks good in this PR but I wasn't able to test the media upload functionality successfully. 

> To test the overall behavior of these changes in action, use this WP Android PR. Once those changes are verified then this PR and the associated PRs can be merged in.

Heads up that we can't use the PR autogenerated APK to test because the feature is DEV only. The branch from the PR was useful for building locally to test though. 

Right now with any type of audio I try and upload from the device, I get an error. My guess is that uploads are failing for me because I need a test site with a premium plan, can you confirm?

image

**Comment by jd-alexander** [comment]
> Hey @jd-alexander . The Android associated code all looks good in this PR but I wasn't able to test the media upload functionality successfully.

Thanks for the review! 

> Heads up that we can't use the PR autogenerated APK to test because the feature is DEV only. The branch from the PR was useful for building locally to test though.

Indeed! Yes, the only way I have seen where I can use the autogenerated APK is to remove the `devOnly` from the audio block but that is a lot of effort. 

> Right now with any type of audio I try and upload from the device, I get an error. My guess is that uploads are failing for me because I need a test site with a premium plan, can you confirm?

Yes, uploads are supposed to fail on the free sites until a fix is created that will prevent the media pickers from selecting MIME types that aren't allowed. Let me add this note to the top of the PR.

**Comment by jd-alexander** [comment]
Resolved a conflict that came in from master ae50b5b Re-running the End-toEnd Tests that failed.

**Review by cameronvoell — APPROVED** [review]
This is looking good from an Android perspective. I was successfully able to upload audio from my device. The media uploader correctly identified audio files on the device to choose from. And after adding audio to the block I was able to update the settings and play the audio while previewing the post. Cool feature!

**Pull request merged**